### PR TITLE
fix semcase on tySequence and tyObject #20283 #19682

### LINKED ISSUE #19682: Using case expression on object result in weird error message
### Example
In this example, the type of X in `case X` is apparent. However, if the X is a complex expression, the error message isn't clear that it is programmer's fault.
```nim
type A = object

case A()
else:
    discard
```

### Current Output
```
test.nim(3, 1) Error: attempting to call undeclared routine: 'case'
```

### Expected Output
A better error message.

### Additional Information

```
Nim Compiler Version 1.7.1 [Linux: amd64]
Compiled at 2022-04-05
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: fa12c59ff0e59e85cf5105bc276d6b151e07c1fa
active boot switches: -d:release
```

### LINKED ISSUE #20283: attempting to call undeclared routine: 'case'
### What happened?

Compiler report `ttempting to call undeclared routine: 'case'` for this code snippet
```nim
case @[]
else: discard
```

### Nim Version

Nim Compiler Version 1.7.1 [Linux: amd64]
Compiled at 2022-08-29
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: ee11302c2493795c630b080fe3dd7ea7721a9fdc
active boot switches: -d:release

### Current Standard Output Logs

```shell
test.nim(1, 1) Error: attempting to call undeclared routine: 'case'
```

### Expected Standard Output Logs

```shell
test.nim(1, 6) Error: selector must be of an ordinal type, float or string
```

### Possible Solution

_No response_

### Additional Information

Versions prior 1.7 work as expected.

### PR DESCRIPTION
**Pull request by bung87**
fix #20283 
fix #19682

Note: I can add a extro param to `resolveOverloads` since this is only one case need to take care of, so that avoiding `if f.kind == nkIdent and f.ident.s == "case":`

edit: now I change to use a flag.  
when I use `case []` , do nim check inside `semStmtList` will cause nil access, so I add a condition `if x == nil`. and <s>the test file produce duplicated error, no idea so far.<s>

### CODE REVIEW

File: compiler/semcall.nim [hunk 1]
```diff
@@ -354,7 +354,8 @@ proc getMsgDiagnostic(c: PContext, flags: TExprFlags, n, f: PNode): string =
 proc resolveOverloads(c: PContext, n, orig: PNode,
                       filter: TSymKinds, flags: TExprFlags,
                       errors: var CandidateErrors,
-                      errorsEnabled: bool): TCandidate =
+                      errorsEnabled: bool,
+                      fastReturn = false): TCandidate =
```
Comment (Araq) [inline compiler/semcall.nim]: Why is this `fastReturn` required?

Comment (bung87) [reply compiler/semcall.nim]: prevent it goes to `localError(c.config, n.info, getMsgDiagnostic(c, flags, n, f))`, ensure not break previouse behavior.

File: compiler/semcall.nim [hunk 2]
```diff
@@ -354,7 +354,8 @@ proc getMsgDiagnostic(c: PContext, flags: TExprFlags, n, f: PNode): string =
 proc resolveOverloads(c: PContext, n, orig: PNode,
                       filter: TSymKinds, flags: TExprFlags,
                       errors: var CandidateErrors,
-                      errorsEnabled: bool): TCandidate =
+                      errorsEnabled: bool,
+                      diagnosticsFlag = true): TCandidate =
```
Comment (Araq) [inline compiler/semcall.nim]: Instead add yet another flag to `TExprFlags` or maybe there is one for this already.

Comment (bung87) [reply compiler/semcall.nim]: `efExplain` ? it use for `errorsEnabled` in `semOverloadedCall`

Comment (bung87) [reply compiler/semcall.nim]: I introduce `efNoDiagnostics` for this.

**Comment by bung87** [comment]
@Araq merge this pls, just a tiny patch

**Pull request merged**